byron-proxy/src/exec/Main.hs

```diff
@@ -401,15 +430,90 @@ runClient tracer clientOptions genesisConfig epochSlots db = case clientOptions
+      _ -> error "no getAddrInfo"
     where
 
     host = scoHostName shelleyClientOptions
     port = scoServiceName shelleyClientOptions
     addrInfoHints = Network.defaultHints
+    -- | This chain sync client will first try to improve the read pointer to
+    -- the tip of the database, and then will roll forward forever, stopping
+    -- if there is a roll-back.
+    -- It makes sense given that we only have an immutable database and one
+    -- source for blocks: one read pointer improve is always enough.
+    chainSyncClient = Client.chainSyncClient fold
+      where
+      fold :: Client.Fold (ResourceT IO) x
+      fold = Client.Fold $ do
+        tip <- lift $ DB.readTip db
+        mPoint <- case tip of
+          -- DB is empty. Can go without improving read pointer.
+          DB.TipGenesis -> pure Nothing
+          -- EBB is nice because we already have the header hash.
+          DB.TipEBB   slotNo hhash _     -> pure $ Just $ ChainSync.Point
+            { ChainSync.pointSlot = slotNo
+            , ChainSync.pointHash = hhash
+            }
+          DB.TipBlock slotNo       bytes -> pure $ Just $ ChainSync.Point
+            { ChainSync.pointSlot = slotNo
+            , ChainSync.pointHash = hhash
+            }
+            where
+            hhash = case Binary.decodeFullAnnotatedBytes "Block or boundary" (Cardano.fromCBORABlockOrBoundary epochSlots) bytes of
+              Left cborError -> error "failed to decode block"
+              Right blk -> case blk of
+                Cardano.ABOBBoundary _ -> error "Corrupt DB: expected block but got EBB"
+                Cardano.ABOBBlock blk  -> Cardano.blockHashAnnotated blk
+        case mPoint of
+          Nothing -> Client.runFold roll
+          -- We don't need to do anything with the result; the point is that
+          -- the server now knows the proper read pointer.
+          Just point -> pure $ Client.Improve [point] $ \_ _ -> roll
+      roll :: Client.Fold (ResourceT IO) x
+      roll = Client.Fold $ pure $ Client.Continue forward backward
+      forward :: ChainSync.Block -> ChainSync.Point -> Client.Fold (ResourceT IO) x
+      forward blk point = Client.Fold $ do
+        lift $ traceWith (contramap chainSyncShow stringTracer) (Right blk, point)
+        -- FIXME
+        -- Write one block at a time. CPS doesn't mix well with the typed
+        -- protocol style.
+        -- This will give terrible performance for the SQLite index as it is
+        -- currently defined. As a workaround, the SQLite index is set to use
+        -- non-synchronous writes (per connection).
+        -- Possible solution: do the batching automatically, within the index
+        -- itself?
+        lift $ DB.appendBlocks db $ \dbAppend ->
+          DB.appendBlock dbAppend (DB.CardanoBlockToWrite blk)
+        Client.runFold roll
+      backward :: ChainSync.Point -> ChainSync.Point -> Client.Fold (ResourceT IO) x
+      backward point1 point2 = Client.Fold $ do
```

Comment:
Is this so simple because we don't expect forks or just because we only have access to immutable DB?

Reply:
Both. I'd deal with forks if I could, but I haven't been given the tools.